spec/12-the-scala-standard-library.md

```diff
@@ -642,6 +642,12 @@ object Predef {
   def classOf[T]: Class[T] = null
    // this is a dummy, classOf is handled by compiler.
 
+  // valueOf -----------------------------------------------------------
+
+  /** Retrieve the single value of a type with a unique inhabitant. */
+  @inline def valueOf[T](implicit vt: ValueOf[T]): T {} = vt.value
```

Comment:
Well.. isn't classOf considered a stable value? (Although it wasn't
originally.) It's pushing it when it takes type arguments, but if it took a
value argument, even implicitly, that would really be stretching it. (Just
being pedantic...)


On Thu, Aug 18, 2016, 8:13 AM Sébastien Doeraene <notifications@github.com>
wrote:

> In spec/12-the-scala-standard-library.md
> <https://github.com/scala/scala/pull/5310#discussion_r75296128>:
>
> > @@ -642,6 +642,12 @@ object Predef {
> >    def classOf[T]: Class[T] = null
> >     // this is a dummy, classOf is handled by compiler.
> >
> > +  // valueOf -----------------------------------------------------------
> > +
> > +  /** Retrieve the single value of a type with a unique inhabitant. */
> > +  @inline def valueOf[T](implicit vt: ValueOf[T]): T {} = vt.value
>
> @soc <https://github.com/soc> I disagree with your reasoning. The fact
> that classOf[T] is a builtin instead of taking an implicit ClassTag[T] is
> an artifact of history. Were we to define classOf and ClassTag *today*,
> we would make the implicit materialization of ClassTag the only builtin,
> and implement classOf in user-space as
>
> def classOf[T](implicit ct: ClassTag[T]): Class[T] = ct.runtimeClass
>
> This would also have a nice benefit that we could simply write classOf[T]
> instead of classTag[T].runtimeClass when we have an implicit ClassTag[T]
> in scope.
>
> Similarly, were new Array[T](len) be defined *today*, it would have an
> implicit ClassTag[T] as well.
>
> Since we are defining ValueOf today, we can do it the right way, with the
> implicit materialization be the only builtin, and valueOf[T] being purely
> user-space.
>
> I do not understand your argument of "bullet-proof". If I have a v:
> ValueOf[T] at my disposal, there's nothing wrong in passing it explicitly
> to valueOf[T](v). The type of v already guarantees that v is the right
> ValueOf[T] to be passed to valueOf[T] anyway.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/5310/files/e82fd54750b8f20c341035cfe086bab0ab0df0e8#r75296128>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGAUCUSt_HXakh1x0X6T982mG7-h7r3ks5qhExagaJpZM4JYDkI>
> .
>